src/openrct2/ride/Ride.cpp

```diff
@@ -3026,19 +3026,98 @@ static void RideOpenBlockBrakes(const CoordsXYE& startElement)
         auto trackType = currentElement.element->AsTrack()->GetTrackType();
         switch (trackType)
         {
+            case TrackElemType::BlockBrakes:
+                BlockBrakeSetLinkedBrakesClosed(
+                    CoordsXYZ(currentElement.x, currentElement.y, currentElement.element->GetBaseZ()),
+                    *currentElement.element->AsTrack(), false);
+                [[fallthrough]];
             case TrackElemType::EndStation:
             case TrackElemType::CableLiftHill:
             case TrackElemType::Up25ToFlat:
             case TrackElemType::Up60ToFlat:
             case TrackElemType::DiagUp25ToFlat:
             case TrackElemType::DiagUp60ToFlat:
-            case TrackElemType::BlockBrakes:
                 currentElement.element->AsTrack()->SetBrakeClosed(false);
                 break;
         }
     } while (track_block_get_next(&currentElement, &currentElement, nullptr, nullptr)
              && currentElement.element != startElement.element);
 }
+void BrakeLinkToBlockBrake(const CoordsXYZ& vehicleTrackLocation, TrackElement& brake)
+{
+    CoordsXYE output = CoordsXYE(vehicleTrackLocation.x, vehicleTrackLocation.y, reinterpret_cast<TileElement*>(&brake));
+    int32_t outputZ = vehicleTrackLocation.z;
+    do
+    {
+        if (output.element->AsTrack()->GetTrackType() == TrackElemType::BlockBrakes)
+        {
+            brake.SetBrakeClosed(
+                (brake.GetBrakeBoosterSpeed() >= output.element->AsTrack()->GetBrakeBoosterSpeed())
+                || (output.element->AsTrack()->IsBrakeClosed()));
+            break;
+        }
+        else if (output.element->AsTrack()->GetTrackType() == TrackElemType::Brakes)
+        {
+            continue;
+        }
+        else
+        {
+            brake.SetBrakeClosed(true);
+            break;
+        }
```

Comment:
If I understand the algorithm correctly, it goes like this:
1. Pick a brake.
2. Search forward
2.1. If we found a block brake, check if we need the close.
2.2. If we found another brake, keep searching until we find a block brake
2.3. If we found another piece, there is no block brake, so always close the brake.
4. Repeat step 1 with another brake until all are done.

Is this correct?

Reply:
1 and 2 yes. This function is only called from TrackSetSpeedAction so it will only affect one TrackElement at a time - ride simulation uses BlockBrakeSetLinkedBrakesClosed.

Note to self: add calls to this and BlockBrakeSetLinkedBrakesClosed in TrackPlaceAction,